Added Admin Warnings For Deleting Conversions

danielshid opened the pull request:
# Description

I added warnings to inform admins of the impact their actions will have on meter units and possible graphing units when choosing to delete conversions.

Partially Addresses #905

## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

I was not able to cover cases for default graphic units and groups.

danielshid committed:
progress on orphan warning

danielshid committed:
implemented check for sources that are meter types

danielshid committed:
logic for suffix meter and count function for redundancy

danielshid committed:
finished implementing impact on meter units and possible graphing units (outline)

danielshid committed:
convert warnings to modals

danielshid committed:
implemented single button for cancelModal

danielshid committed:
Merge branch 'development' into issue#905

danielshid committed:
fix conditional

danielshid committed:
changed warnings for singular and plural

huss commented on src/client/app/components/conversion/EditConversionModalComponent.tsx:
```diff
@@ -59,10 +63,80 @@ export default function EditConversionModalComponent(props: EditConversionModalC
 	};
 	/* End State */
 
+	const getConversionCount = (unit: UnitData, conversions: ConversionData[]) => {
```
Each function should have a comment (similar to JSDoc) to tell what it does and the params. It would also be nice to have comments inside the functions to tell what is being done in the big picture sense.

huss commented on src/client/app/components/conversion/EditConversionModalComponent.tsx:
```diff
@@ -59,10 +63,80 @@ export default function EditConversionModalComponent(props: EditConversionModalC
 
+	const getConversionCount = (unit: UnitData, conversions: ConversionData[]) => {
+		let count = 0;
+		if (unit === unitDataById[state.sourceId]) {
```
I think these types of compare of entire unit objects work but OED more commonly uses the .id for the check.

huss commented on src/client/app/components/conversion/EditConversionModalComponent.tsx:
```diff
@@ -59,10 +63,80 @@ export default function EditConversionModalComponent(props: EditConversionModalC
+			} else {
+				msg += 'Deleting this meter conversion will make the following meter(s) ungraphable:\n';
+				for (const meterId of Object.values(meterDataById)) {
+					if (meterId.unitId == source.id) {
```
Should this be ===?

huss commented on src/client/app/components/conversion/EditConversionModalComponent.tsx:
```diff
@@ -59,10 +63,80 @@ export default function EditConversionModalComponent(props: EditConversionModalC
+		if (source.typeOfUnit === UnitType.meter) {
+			const count = getConversionCount(source, conversionDetails);
+			if (count === 1) {
+				msg += `Deleting this meter conversion will orphan ${unitDataById[state.destinationId].name}.\n`;
```
All strings should be translated.

huss commented on src/client/app/components/conversion/EditConversionModalComponent.tsx:
```diff
@@ -59,10 +63,80 @@ export default function EditConversionModalComponent(props: EditConversionModalC
+			if (count === 1) {
+				msg += `Deleting this meter conversion will orphan ${unitDataById[state.destinationId].name}.\n`;
+			} else {
+				msg += 'Deleting this meter conversion will make the following meter(s) ungraphable:\n';
```
I'm not sure the logic is what was desired. If there is only this conversion and there are meter(s) using the source unit then tell the admin about each meter and cancel the delete. If there are multiple conversions then tell the admin about each meter using the source unit and warn that the number of graphable units for the meter is reduced.

huss commented on src/client/app/components/conversion/EditConversionModalComponent.tsx:
```diff
@@ -59,10 +63,80 @@ export default function EditConversionModalComponent(props: EditConversionModalC
+			if (destCount === 1) {
+				msg += `Deleting this unit conversion will orphan ${unitDataById[state.destinationId].name}.\n`;
+			}
+			if (state.bidirectional) {
```
I think the description of the cases I gave you was confusing. What needs to be determined (including the check right above) is if there is only this conversion for the unit. To check this, the code needs to check if that unit is a source or if the conversion is bidirectional then it can be the destination (excluding the one being deleted). The idea is if it is the destination then it can could get to another unit if allowed which means it must be bidirectional. This check is looking at the bidirectional status of the one being deleted and not the other possibilities.

huss commented on src/client/app/components/conversion/EditConversionModalComponent.tsx:
```diff
@@ -59,10 +63,80 @@ export default function EditConversionModalComponent(props: EditConversionModalC
+				}
+			}
+			if (msg === '') {
+				msg += 'Deleting this unit conversion between two units of type unit will have consequences.\n';
```
I think change "will" to "may" since it can happen. Overall, I think the message might be better as: "Deleting this conversion between two units of type unit may reduce the possible graphing units for some meter(s) but cannot be known until after the deletion happens." I also would like a TODO put here to indicate that OED should try to add a check after the deletion to see if a change happens to notify the admin of the consequence of this delete. In a perfect world the admin could ask to have the delete undone.

huss commented on src/client/app/components/conversion/EditConversionModalComponent.tsx:
```diff
@@ -59,10 +63,80 @@ export default function EditConversionModalComponent(props: EditConversionModalC
+		if (msg === '') {
+			handleDeleteConfirmationModalOpen();
+		} else if (cancel) {
+			setDeleteConfirmationMessage(msg + 'This conversion cannot be deleted.\n');
```
Could the new part of the msg be: "This conversion cannot be deleted until the significant consequences of doing this are addressed."

huss commented on src/client/app/styles/modal.css:
```diff
@@ -17,4 +17,8 @@ select option[value="false"] {
+
+.confirmation-message {
+	white-space: pre-wrap;
+  }
```
Please remove extra spaces at start of line.

huss commented on src/client/app/components/conversion/EditConversionModalComponent.tsx:
```diff
@@ -59,10 +63,80 @@ export default function EditConversionModalComponent(props: EditConversionModalC
+		} else if (source.typeOfUnit === UnitType.unit && dest.typeOfUnit === UnitType.unit) {
+			const destCount = getConversionCount(dest, conversionDetails);
+			if (destCount === 1) {
+				msg += `Deleting this unit conversion will orphan ${unitDataById[state.destinationId].name}.\n`;
```
Could we reword so it has "will orphan unit <unit name>" (so adding word unit). Also, generally, it would be nice if there were quotes around the name in all messages since unit names can have spaces and multiple words.

huss reviewed (changes requested):
Thanks to @danielshid for working on this issue. Overall, it is fine. I have made some comments about details and style for you to consider. Given the cases are complex, this is not surprising. I did some testing and it works as desired (modulo the comments). I'll do a final check after the comments are addressed. Let me know if I can help in any way.

danielshid committed:
turned strings in checkState into translations

danielshid committed:
revised getConversionCount() and corrected checkState() if source is meter

danielshid committed:
changed action when deleting conversion involving two units of type unit

huss committed:
fix up delete conversion checks

- The issue did not correctly describe all the needed
conditions and it is believe this fixes the code so it
works as needed.
- Suffix units has issues so it is commented out with
a TODO for future work.
- Minor wording update to user messages.

huss commented:
I've modified the code to fix several issues that careful testing found with the design given in the issue on GitHub. This was my fault and not in the code provided. I'm going to let this sit a little while in case anyone else (including @danielshid who did this PR) has thoughts on the modifications.

huss reviewed (approved):
Thanks to @danielshid for continued work on this after the time expected. I'm sorry this sat for a couple of weeks while I was focused elsewhere.

All your changes looked good and worked as requested. Please see my comment on a commit I made to update the behavior that is outside what was requested.

I'm approving the work but will wait to merge per my other comment. Congratulations to @danielshid for another successful contribution to OED.
